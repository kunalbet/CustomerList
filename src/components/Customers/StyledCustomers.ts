import styled from "styled-components";
import TableRow from '@material-ui/core/TableRow';
import TableCell from '@material-ui/core/TableCell';

export const StyledTableRow = styled(TableRow)`
    background-color: lightskyblue;
`;

export const StyledTableCell = styled(TableCell)`
    color: white;
`;

export const StyleTableHeader = {
    color: 'white',
    'font-size': '16px'
};