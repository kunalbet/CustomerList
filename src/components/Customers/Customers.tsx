import * as React from "react";
import { Dispatch } from "redux";
import { useDispatch } from "react-redux";
import { ICustomer } from "../../types/types";
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import TableRow from '@material-ui/core/TableRow';
import TableCell from '@material-ui/core/TableCell';
import { StyledTableRow } from './StyledCustomers';

type Props = {
  customers: readonly ICustomer[];
  removeCustomer: (customer: ICustomer) => void;
};

export const Customers: React.FC<Props> = ({ customers, removeCustomer }) => {
  const dispatch: Dispatch<any> = useDispatch();

  const deleteCustomer = React.useCallback(
    (customer: ICustomer) => dispatch(removeCustomer(customer)),
    [dispatch, removeCustomer]
  );

  return (
      <TableContainer component={Paper}>
        <Table>
          <TableHead>
            <StyledTableRow>
              <TableCell><b>Sr.No</b></TableCell>
              <TableCell align="right"><b>First Name</b></TableCell>
              <TableCell align="right"><b>Last Name</b></TableCell>
              <TableCell align="right"><b>Phone Number</b></TableCell>
              <TableCell align="right">&nbsp;</TableCell>
            </StyledTableRow>
          </TableHead>
          <TableBody>
            {customers.map((customer: ICustomer, index) => (
              <TableRow key={customer.id} data-testid="customer-row">
                <TableCell component="th" scope="row">
                  {index + 1}
                </TableCell>
                <TableCell align="right">{customer.firstName}</TableCell>
                <TableCell align="right">{customer.lastName}</TableCell>
                <TableCell align="right">{customer.phoneNumber}</TableCell>
                <TableCell align="right">
                  <Button variant="contained" color="secondary" data-testid="delete-button"
                  onClick={() => deleteCustomer(customer)}>Delete</Button>
                </TableCell>
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </TableContainer>
  );
};
