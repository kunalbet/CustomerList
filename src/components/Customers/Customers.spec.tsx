import * as React from "react";
import { render, screen, fireEvent } from "../../utils/testUtils";
import { Customers } from "./Customers";
import * as redux from "react-redux"; 

const customers = [{
  id: 1,
  firstName: "Test",
  lastName: "Tester",
  phoneNumber: "00000000",
}];

describe("<Customers />", () => {
  it("should render a <Customers />", () => {
    const wrapper = render(
      <Customers customers={customers} removeCustomer={() => {}} />
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('renders customers table', () => {
    render(<Customers customers={customers} removeCustomer={() => {}} />);
    const linkElement = screen.getByText(/First Name/);
    const linkElement1 = screen.getByText(/Last Name/);
    expect(linkElement).toBeInTheDocument();
    expect(linkElement1).toBeInTheDocument();
  });

  it('renders customers Data in table', () => {
    render(<Customers customers={customers} removeCustomer={() => {}} />);
    const customerRows = screen.getAllByTestId('customer-row');
    expect(customerRows.length).toBe(1);
  });

  it('calls removeCustomer prop with currect customer when DELETE button is clicked', () => {
    const mockRemoveCustomer = jest.fn();
    const useDispatchSpy = jest.spyOn(redux, 'useDispatch'); 
    const mockDispatchFn = jest.fn();
    useDispatchSpy.mockReturnValue(mockDispatchFn);
    render(<Customers customers={customers} removeCustomer={mockRemoveCustomer} />);
    fireEvent.click(screen.getByTestId('delete-button'));
    expect(mockDispatchFn).toHaveBeenCalledTimes(1);
    expect(mockRemoveCustomer).toHaveBeenCalledWith(customers[0]);
  })
});
