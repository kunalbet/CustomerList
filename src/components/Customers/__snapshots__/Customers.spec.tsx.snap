// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Customers /> should render a <Customers /> 1`] = `
<div>
  <div
    class="MuiPaper-root MuiTableContainer-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <table
      class="MuiTable-root"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root sc-bdfBQB bYKAzC MuiTableRow-head"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            <b>
              Sr.No
            </b>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            <b>
              First Name
            </b>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            <b>
              Last Name
            </b>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            <b>
              Phone Number
            </b>
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
            scope="col"
          >
            Â 
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root"
          data-testid="customer-row"
        >
          <th
            class="MuiTableCell-root MuiTableCell-body"
            role="cell"
            scope="row"
          >
            1
          </th>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
          >
            Test
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
          >
            Tester
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
          >
            00000000
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
              data-testid="delete-button"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label"
              >
                Delete
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
